src/phoenix/core/traces.py

```diff
@@ -1,108 +1,267 @@
-        self._adjacency_lists = defaultdict(list)
-        for span_id, parent_id in cast(
-            Iterable[Tuple[SpanID, SpanID]],
-            parent_span_ids.dropna().items(),
-        ):
-            self._adjacency_lists[parent_id].append(span_id)
-
-    def get_descendant_span_ids(self, span_id: SpanID) -> Iterable[SpanID]:
-        for span_id in self._adjacency_lists[span_id]:
-            yield span_id
-            yield from self.get_descendant_span_ids(span_id)
-
-
-def _cumulative(
-    span_values: "pd.Series[Any]",
-    span_parent_ids: "pd.Series[Any]",
-) -> "pd.Series[Any]":
-    """
-    For each span, return the cumulative (i.e. summed) value from itself and
-    all of its descendants (i.e. children, grandchildren, etc.).
-
-    Example
-    -------
-    >>> span_ids = list("ABCDEF")
-    >>> span_values = pd.Series([1,1,None,1,1,None], index=span_ids)
-    >>> span_parent_ids = pd.Series([None,"A","A","C","C",None], index=span_ids)
-    >>> _cumulative(span_values, span_parent_ids)
-    A    4.0
-    B    1.0
-    C    2.0
-    D    1.0
-    E    1.0
-    F    0.0
-    dtype: float64
-
-    Parameters
-    ----------
-    span_values: pd.Series, indexed by span_id
-        each span's numeric value to be accumulated
-    span_parent_ids: pd.Series, indexed by span_id, same shape as `values`
-        each span's parent span_id (may be null on some rows)
-
-    Returns
-    -------
-    accumulated_values: pd.Series, indexed by span_id, same shape as `values`
-        each span's cumulative value (from itself and all of its descendants)
-    """
-    cumulative_values = pd.Series(
-        np.zeros_like(span_values),
-        index=span_values.index,
-    )
-    for span_id, value in span_values.items():
-        if not value or isinstance(value, float) and math.isnan(value):
-            continue
-        while span_id:
-            cumulative_values.loc[span_id] += value  # type: ignore
-            span_id = span_parent_ids.loc[span_id]  # type: ignore
-    return cumulative_values
+            existing_value = (existing_span[attribute_name] or 0) if existing_span else 0
+            new_value = new_span[attribute_name] or 0
+            if not (difference := new_value - existing_value):
+                continue
+            existing_cumulative_value = (
+                (existing_span[cumulative_attribute_name] or 0) if existing_span else 0
+            )
+            new_span[cumulative_attribute_name] = difference + existing_cumulative_value
+            self._add_value_to_span_ancestors(
+                difference,
+                span_id,
+                cumulative_attribute_name,
+            )
+        # Process previously orphaned spans, if any.
+        for orphan_span in self._orphan_spans[span_id]:
+            self._process_span(orphan_span)
+
+    def _add_value_to_span_ancestors(
+        self,
+        value: float,
+        span_id: SpanID,
+        cumulative_attribute_name: str,
+    ) -> None:
```

Comment:
the function signature is a bit complex and in a strange order?
```suggestion
    def _add_value_to_span_ancestors(
        self,
        span_id: SpanID,
        attribute_name: str,
        value: float,
    ) -> None:
```